File: src/main/scala/com/nibado/projects/adventscala/Day07.scala
package com.nibado.projects.adventscala

import scala.collection.mutable

class Day07 extends Day {
  val gatePattern = "^((?<w1>[0-9a-z]{1,3}) )?((?<gate>AND|OR|NOT|RSHIFT|LSHIFT) )?(?<w2>[0-9a-z]{1,5}) -> (?<wO>[a-z]{1,2})$".r.pattern

  override def run(): Unit = {
    val input = getResource("/day07.txt")

    val network1 = createNetwork(input)
    val network2 = createNetwork(input)

    val result1 = network1.get("a").get.solve()

    printAnswer(7, "One", result1)
    network2.get("b").get.value = result1
    printAnswer(7, "Two", network2.get("a").get.solve())
  }

  def createNetwork(input: List[String]): mutable.HashMap[String, Wire] = {
    val map = new mutable.HashMap[String, Wire]()

    def getWire(id: String): Option[Wire] = {
      if(id == null) {
        return null
      }
      if(id.matches("[0-9]{1,5}")) {
        val wire = new Wire()
        wire.value = id.toInt

        return Option(wire)
      }

      if(!map.contains(id)) {
        map.put(id, new Wire())
      }

      map.get(id)
    }

    input.map(s => gatePattern.matcher(s)).filter(m => m.matches()).foreach(m => {
      val wireOut: Wire = getWire(m.group("wO")).get

      wireOut.gate = new Gate(m.group("gate"), getWire(m.group("w1")), getWire(m.group("w2")))
    })

    map
  }

   class Wire {
    var value: Int = -1
    var gate: Gate = null

    def solve() : Int = {
      if(value < 0) {
        value = gate.solve()
      }
      value
    }
  }

  class Gate(val gate: String, val wire1: Option[Wire], val wire2: Option[Wire]) {
    def solve(): Int =
      gate match {
        case null => wire2.get.solve()
        case "AND" => wire1.get.solve() & wire2.get.solve()
        case "OR" => wire1.get.solve() | wire2.get.solve()
        case "NOT" => (~wire2.get.solve()) & 0xFFFF
        case "LSHIFT" => wire1.get.solve() << wire2.get.solve()
        case "RSHIFT" => wire1.get.solve() >> wire2.get.solve()
        case _ => throw new RuntimeException("Unknown gate: " + gate)
      }
  }
}

File: src/main/java/com/nibado/projects/advent/Day07.java
package com.nibado.projects.advent;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static com.nibado.projects.advent.Util.printAnswer;
import static com.nibado.projects.advent.Util.readResource;


public class Day07 implements Runnable {
    private static final Pattern GATE = Pattern.compile("^((?<w1>[0-9a-z]{1,3}) )?((?<gate>AND|OR|NOT|RSHIFT|LSHIFT) )?(?<w2>[0-9a-z]{1,5}) -> (?<wO>[a-z]{1,2})$");
    private static final Pattern LITERAL = Pattern.compile("[0-9]{1,5}");

    public static void main(String... argv) {
        new Day07().run();
    }

    @Override
    public void run() {
        List<String> input = readResource("/day07.txt");

        GateFactory factory = new GateFactory();
        input.stream().forEach(factory::add);

        int value = factory.get("a").solve();
        printAnswer(value);

        factory = new GateFactory();
        input.stream().forEach(factory::add);

        factory.get("b").value = value;
        printAnswer(factory.get("a").solve());
    }

    public static class GateFactory {
        private Map<String, Wire> wires = new HashMap<>();

        public Wire get(String wireId) {
            if(wireId == null) {
                return null;
            }
            else if(LITERAL.matcher(wireId).matches()) {
                Wire w = new Wire();
                w.value = Integer.parseInt(wireId);

                return w;
            }
            if(!wires.containsKey(wireId)) {
                wires.put(wireId, new Wire());
            }

            return wires.get(wireId);
        }

        public void add(String line) {
            Matcher m = GATE.matcher(line);
            if(!m.matches()) {
                throw new IllegalArgumentException(line + " doesn't match gate regex");
            }

            Wire wireOut = get(m.group("wO"));

            wireOut.gate = new Gate();
            wireOut.gate.type = m.group("gate");
            wireOut.gate.wire1 = get(m.group("w1"));
            wireOut.gate.wire2 = get(m.group("w2"));
        }
    }

    public static class Wire {
        private int value = -1;
        private Gate gate = null;

        public int solve() {
            if(value < 0) {
                value = gate.solve();
            }

            return value;
        }
    }

    public static class Gate {
        private String type;
        private Wire wire1;
        private Wire wire2;

        public int solve() {
            if(type == null) {
                return wire2.solve();
            }
            switch(type) {
                case "AND": return wire1.solve() & wire2.solve();
                case "OR": return wire1.solve() | wire2.solve();
                case "NOT": return ~wire2.solve() & 0xFFFF;
                case "LSHIFT": return wire1.solve() << wire2.solve();
                case "RSHIFT": return wire1.solve() >> wire2.solve();
                default: throw new IllegalArgumentException("Unknown gate: " + type);
            }
        }
    }
}
